### PR TITLE
Typed user account API: display name, email updates, extra preferences, address gating

### PR DESCRIPTION
**Pull request by itisAliRH**
Regroups the server-side half of the user account work that was previously split across this PR and #23304, #23305 and #23308. Same commits, same content, one PR against `dev` instead of four stacked branches — per the review feedback that the split was too atomic.

The client half (the native Account page and the preference grouping, previously #23306 and #23307) follows in a separate PR once this one lands. The migration data-fix this used to sit on top of is #23314, merged on its own first: adding a column to `galaxy_user` is exactly what the old data fix selecting the whole ORM model broke on.

---

## Display name on `galaxy_user`

Adds a `display_name` column to `galaxy_user`: a free-form name shown in place of the username.

Galaxy has one name per user doing three jobs at once — it signs you in, it is the `/u/{username}/...` URL segment, and it is what other people see. Those pull in different directions: a URL segment wants to be short, unique and lowercase; a name people see wants to be whatever the person is actually called. This adds the presentational half so the username can stop pretending to be both. It is groundwork for the public profile pages work, but it is useful on its own — it is what lets a later change show a real name next to published items.

**Why on `galaxy_user` rather than a profile table.** The profile pages branch puts profile fields on a `user_profile` side table gated behind `enable_user_profile_pages`. A display name has to work on every instance, so it belongs on the user.

Design notes:

- Nullable, falling back to the username. Nothing is backfilled — making it required would mean either copying every username into it or blocking every account save on an empty field.
- Not unique and not indexed. Display names collide by design, nothing resolves a user by one, and the only plausible consumer (admin user search) uses a leading-wildcard `LIKE` that an index cannot serve.
- Exposed on `DetailedUserModel` only, deliberately not on `BaseUserModel`, `UserModel` or the `Dictifiable` key lists. Those feed `GET /api/users`, and an instance setting `expose_user_name: false` has chosen to hide identity; a real name leaking through that gap would be worse than the username it replaces.
- One writer, `PUT /api/users/{user_id}`. Two write paths would mean two validation sites and undefined precedence.
- Validation is a denylist, not an allowlist: names are free-form and may be in any script, so the rule is length plus the invisible characters (C0/C1 controls, zero-width, bidirectional formatting). The bidi overrides are why the check exists — `U+202E` lets a stored name render as text unrelated to what is saved, which turns a non-unique field into an impersonation vector.
- Deliberately **not** doing: rejecting names that collide with an existing username (a query per save, a nonsensical "taken" error on a non-unique field, trivially bypassed), or an `admin`/`galaxy` denylist (blocks a real given name and a real organisation while `Galaxy Admin` and `GaIaxy` sail through). The anti-impersonation control is that the `@username` stays visible wherever a display name renders.

`UserManager.purge` hashed email and username but would have left `display_name` — free text that typically holds a real name — on the row. It is now dropped on purge; dropped rather than hashed, because unlike the username there is no reason to re-derive it.

Anyone who has already run this branch's migration on a dev database gets the column added normally; the migration is a plain nullable `add_column` and round-trips cleanly.

---

## Email changes via `PUT /api/users/{user_id}`

`PUT /api/users/{user_id}` now accepts `email`.

Changing an email address was previously only possible through `PUT /api/users/{id}/information/inputs`, a legacy `@expose_api` WSGI action that is absent from the OpenAPI schema and therefore invisible to the typed client. A typed endpoint for the same resource already existed and already handled `username`, so this closes the gap rather than adding a third way to write to a user.

The deserializer delegates to `UserManager.update_email(..., commit=False, send_activation_email=True)` — the identical call the legacy handler makes — so the following are unchanged: validation including duplicate addresses and blocked or non-allowlisted domains; the private role rename (the role is named for the email and is what dataset permissions are granted against); and deactivate-and-mail when `user_activation_on` is set. `ModelDeserializer.deserialize` supplies the single trailing commit, which is what lets `update_email` roll back when the activation mail fails to send.

`active` on the same endpoint is now admin-only. It was a bare `setattr` for any authenticated user, which with `user_activation_on` let someone change their email, get deactivated, and then re-activate themselves without ever verifying the new address. That hole predates this PR; the gate is its own commit so it can be backported to the release branches. A non-admin sending `active` gets a 403, and the field description says so. The only client that sends `active` is the admin user grid.

Out of scope, but worth recording: changing an email requires no re-authentication, while changing a password requires the current password. That is pre-existing and unchanged here.

---

## Typed endpoints for extra user preferences

Adds `GET|PUT /api/users/{user_id}/extra_preferences/inputs` and moves the admin-defined extra-preference handling out of the legacy controller into `UsersService`.

`information/inputs` was doing two unrelated jobs: a fixed, typed set of account fields (email, username), and an open-ended set of admin-authored sections from `user_preferences_extra_conf.yml`. The first belongs on the typed user endpoint. The second is genuinely dynamic — its shape comes from an administrator's YAML — so form-JSON and the generic form remain the right tools for it; it just needs its own address rather than being appended to the account page.

The logic **moves**, it is not copied: `information/inputs` now delegates to the same service methods, so the deprecated endpoint and the new one cannot drift apart and every existing caller keeps working. The vault and `__SECRET_PLACEHOLDER__` handling is moved verbatim — it is fiddly, its only coverage needs a real vault, and a regression there would silently overwrite users' stored secrets with the literal placeholder string, so it is deliberately not tidied up on the way past.

Also adds a derived `has_user_preferences_extra` config flag, mirroring `has_user_tool_filters`, so the client can hide the entry on instances that configure nothing rather than linking to an empty form.

`information/inputs` now logs a deprecation warning once per process and carries a `.. deprecated::` note. It should be removed a release from now, together with the `buildapp.py` mapper entries. One loose end for that removal: the `info` conditional (admin `FormDefinition` USER_INFO forms) is the only part of the endpoint with no new home — it is admin-configured and dynamic, the same category as extra preferences, and it would be worth asking the admin community whether any deployment still uses it.

---

## Postal addresses behind `enable_user_addresses`

Adds `enable_user_addresses` (default `true`) gating postal address storage, and deletes two dead helpers.

`UserAddress` is a leftover. Nothing in Galaxy reads these addresses — they are written and read by exactly one place, the deprecated `information/inputs` endpoint. There is no manager, job, tool or quota that consults them. Galaxy's interface no longer offers them at all now that the Account page has replaced the generic form, so this option only affects that deprecated endpoint. It gives administrators a way to stop accepting addresses ahead of the table being dropped, and gives that removal a single line to flip.

The default is `true`, so nothing changes for API consumers today.

`save_widget_field` and `get_form_values` in `webapps/base/controller.py` have no callers anywhere in `lib` or `test`. `get_form_values` was the only code path that could create a `UserAddress` from a generic form, so removing it also removes the last writer outside the deprecated endpoint.

One detail worth a reviewer's eye: the gate wraps the **whole** address block, not just the payload parsing. The block rebuilds `user.addresses` from scratch on every save, so gating only the parsing would have left `user.addresses = []` running unconditionally, silently discarding stored addresses the first time anyone saved with the feature off.

Dropping the table itself needs team sign-off and is irreversible — it would remove `user_address`, `AddressField`, the `address` form-builder type, the GDPR redaction branch, and this option. It is blocked on `test_tools.py::test_model_attributes_sanitization`, which uses an address field as its vehicle for a string containing a double quote and asserts on the `model_attributes` tool's output; that needs a different field first.

---

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. `./run_tests.sh -unit test/unit/data/security/test_validate_user_input.py test/unit/app/managers/test_UserManager.py test/unit/data/model/db/test_user.py`
  2. `./run_tests.sh -api lib/galaxy_test/api/test_users.py`
  3. `./run_tests.sh -integration test/integration/test_users.py test/integration/test_vault_extra_prefs.py`
  4. Migration round trip: `sh manage_db.sh upgrade gxy@head`, confirm `galaxy_user.display_name` exists, then `sh manage_db.sh downgrade <previous>` and confirm it is gone. `sh manage_db.sh heads` should still show exactly one `gxy` head.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by itisAliRH** [comment]
CI after the rebase onto `dev` (#23314 is in): the `migration_fixes` unit tests that were the only real red are green now. The four remaining failures are unrelated to this branch:

- `Test Galaxy packages` (3.10 and 3.14): `test_bam.py::test_merge_bam` fails with `samtools merge: failed to get @HD line from header` / a vanished temp dir. Every packages run today fails the same way, including unrelated PRs (e.g. `sniff-pgm-pbm`), so this is the runner environment.
- `Selenium tests`: `test_workflow_editor` / `test_workflow_run` sample sheet chipseq cases and `test_history_storage::test_delete_dataset_from_storage_view`. The sample sheet editor test also fails on the latest `dev` run (c3c83bec99); none of these touch the user account code.

**Comment by jmchilton** [comment]
Here are some headline issues from my Claude review - double checked with a subagent. Can you address them or pushback if they are slop. They read well to me but may be non-sense. There is some more reuse stuff but let's see if this is valid or non-sense first.

<details>

### B1. `UserManager.update_display_name` is dead code; the deserializer duplicates it

*Correctness-neutral — nothing misbehaves at runtime. Blocking on reuse/dead-code grounds:
the PR ships an unreachable method and a duplicated validation rule on a surface it is
explicitly typing.*

`lib/galaxy/managers/users.py:227-244` adds:

```python
def update_display_name(self, user: User, new_display_name: str | None, *, commit: bool = True) -> None:
    normalized = (new_display_name or "").strip() or None
    if message := validate_display_name_str(normalized):
        raise exceptions.RequestParameterInvalidException(message)
    ...
```

`lib/galaxy/managers/users.py:869-873` adds, separately:

```python
def deserialize_display_name(self, item, key, display_name, **context):
    display_name = (display_name or "").strip() or None
    if validation_error := validate_display_name_str(display_name):
        raise base.ModelDeserializingError(validation_error)
    return self.default_deserializer(item, key, display_name, **context)
```

`grep -rn update_display_name lib/ test/ client/src` returns **only the definition and unit
tests** — no production caller, and no dynamic dispatch reaches it: `add_deserializers`
(`managers/users.py:842-847`) maps `"display_name"` to `self.deserialize_display_name`, and
`ModelDeserializer.deserialize` (`managers/base.py:920`) only ever calls `self.deserializers[key]`,
never a manager method by name. So the manager method is dead, the strip-and-validate rule exists twice, and
the manager unit tests (`test/unit/app/managers/test_UserManager.py:316-359`) are exercising a
code path no request reaches. The single API-reachable path is the deserializer, tested
separately at `test/unit/app/managers/test_UserManager.py:528-543`.

This is the reuse concern, and it is sharper because **the same PR does it correctly for
email**: `deserialize_email` (`managers/users.py:849-856`) calls
`self.manager.update_email(..., commit=False)`. `deserialize_display_name` should call
`self.manager.update_display_name(item, display_name, commit=False)` the same way — one
validation site, the manager as the seam, and the manager tests then cover the real path.

(Note the existing `deserialize_username` also bypasses `update_username` — that's pre-existing,
but it means this PR had a choice of two precedents and picked the older, worse one for
display_name while picking the better one for email.)

The PR description's design note — *"One writer, `PUT /api/users/{user_id}`. Two write paths
would mean two validation sites and undefined precedence."* — describes the intent, but the code
as merged has two validation sites anyway, one of them unreachable.

### B2. `PUT /api/users/{user_id}` with `{"email": null}` is a 500

`UserUpdatePayload.email` is `str | None` with default `None` (`lib/galaxy/schema/schema.py:424-436`),
so `{"email": null}` passes schema validation, and `payload.model_dump(exclude_unset=True)`
(`lib/galaxy/webapps/galaxy/api/users.py:719`) **includes** an explicitly-set null.
`deserialize_email` passes it straight through with no coercion:

```python
self.manager.update_email(trans, item, email, commit=False, send_activation_email=True)
```

`update_email` (`lib/galaxy/managers/users.py:196`) calls `validate_email(trans, None, user)`,
and in `lib/galaxy/security/validate_user_input.py` the duplicate-check statement is **not**
guarded by the earlier `message` short-circuit:

```python
    if not message:
        if is_email_banned(...):
            message = "This email address has been banned."

    stmt = select(trans.app.model.User).filter(func.lower(trans.app.model.User.email) == email.lower()).limit(1)
```

`None.lower()` → `AttributeError` → 500 rather than 400.

**This crash is pre-existing, not introduced here.** At the merge-base, `set_information`
(`api/users.py:947-955`) already did `if "email" in payload: update_email(trans, user,
payload.get("email"))`, so `{"email": null}` 500s on `dev` today via `information/inputs`. What is
new is that the typed schema now *advertises* `email: str | null` as an accepted input on a
documented endpoint — a client can reasonably send it from the OpenAPI spec. Fixing it is still
this PR's job, but frame it to the author as "the typed endpoint documents an input that crashes",
not "you added a crash".

Contrast `deserialize_display_name`, which does coerce (`(display_name or "").strip() or None`).
`deserialize_email` should reject/normalise `None` before calling the manager (or `validate_email`
should return early on a falsy address — that also fixes the same latent 500 on the legacy
`information/inputs` path, `api/users.py:952-961`).

Confidence: **confirmed by inspection**, not by execution (no test suites run per instructions).
`email: Annotated[str | None, ...] = None` accepts `null` with no 422 at the schema boundary; a
standalone pydantic check confirms `model_validate({"email": None}).model_dump(exclude_unset=True)`
→ `{'email': None}`; and the `stmt = select(...)` line sits at function indent, outside the
`if not message:` guards. Still worth a one-line API test before posting if you want it stated
flatly rather than as "believe this 500s".

---

## Should-fix

### S1. The PR description's claim about an "admin gate on `active`" is false

`lib/galaxy/schema/schema.py:426-428` comments the `email` field:

```python
    # Declared last so that a payload combining it with `active` ends on the
    # deactivation, not on a stale activation. The admin gate on `active` is the
    # real protection; this is belt and braces.
```

There is no admin gate on `active`. `UserDeserializer.add_deserializers`
(`lib/galaxy/managers/users.py:842`) registers `"active": self.default_deserializer`, and the
endpoint (`lib/galaxy/webapps/galaxy/api/users.py:709-721`) has no `require_admin=True`; its only
check is `get_non_anonymous_user_full`, which permits self.

So with `user_activation_on` set, a logged-in non-admin can:

1. `PUT /api/users/{self}` `{"email": "new@x"}` → deactivated, activation mail sent.
2. `PUT /api/users/{self}` `{"active": true}` → reactivated, new address never verified.

Step 2 works because activation is enforced **only at job-execution points** —
`trans.check_user_activation()` appears in `tools/actions/*`, `services/workflows.py:185` and
`services/wes.py:477`, and nowhere in the users API. The deactivated session stays valid, so the
user can turn `active` back on themselves.

The declaration-order trick only holds within a single payload; it does nothing across two
requests.

**The exploit chain is entirely pre-existing.** `active` was already on `UserUpdatePayload` at the
merge-base (`git show <merge-base>:lib/galaxy/schema/schema.py`, line 411) with the same bare
`setattr` deserializer, and the merge-base legacy PUT already let a user change their own email.
Neither half is new. **The only defect this PR introduces is the comment** — it documents a
protection that has never existed, which is worse than silence because the next reader will trust
it. Either gate `active` on `trans.user_is_admin` in the deserializer (a real fix, arguably out of
scope), or correct the comment to say plainly that self-reactivation is an open pre-existing hole.
Do not let it ship asserting a gate.

</details>

**Comment by itisAliRH** [comment]
All three hold up; fixed in the last commits, nothing to push back on, thanks, @jmchilton.

**B1** — `deserialize_display_name` now delegates to `UserManager.update_display_name(item, display_name, commit=False)`, the same shape as `deserialize_email`, so there is one strip-and-validate site and the manager tests exercise the path a request actually takes (d348bc5f52). `deserialize_username` still bypasses `update_username`; that is pre-existing and left alone here, happy to do it as a follow-up.

**B2** — Reproduced rather than inspected: `validate_email(trans, None)` raises `AttributeError: 'NoneType' object has no attribute 'lower'`, because the duplicate-address `select` is built outside the `if not message` guard. Moved it under the guard, so `{"email": null}` now gets the 400 "No email address was provided." on both the typed endpoint and the legacy `information/inputs` path. API test added (59ba4551cd).

**S1** — Correct, the comment described a gate that did not exist. There is one now: `UserDeserializer` routes `active` through `deserialize_active`, which raises 403 for anyone who is not an administrator; the schema comment and the field description now say what the code does. The only client caller that sends `active` is the admin user grid, so nothing else changes. Tests cover a non-admin trying to re-activate themselves (403) and an admin toggling it (200) or sending a null (400) (1724d37a8c, client schema regen in be3ecb47bb). The self-reactivation chain is pre-existing on `dev` and the release branches; the fix commit stands alone, so it can be backported to `release_26.1` and `release_26.0` if you want that.

Send the reuse items whenever.

File: lib/galaxy_test/api/test_users.py [hunk 1]
```diff
@@ -112,6 +114,35 @@ def test_update(self):
             update_response = self._put(update_url, data=payload, json=True)
             assert_object_id_error(update_response)
 
+    @requires_new_user
+    def test_update_display_name(self):
+        # This suite runs with public profile pages disabled, which is the point:
+        # a display name has to be settable on every instance, not only on the
+        # ones that expose profile pages.
```
Comment (jmchilton) [inline lib/galaxy_test/api/test_users.py:121]: This isn't a configuration dependent thing - this comment is confusing and wrong I think but the test is fine.

File: test/unit/app/managers/test_UserManager.py [hunk 2]
```diff
@@ -313,6 +313,52 @@ def test_reset_email_user_deleted(self):
         message = self.user_manager.send_reset_email(cast("GalaxyWebTransaction", self.trans), {"email": user_email})
         assert message is None
 
+    def test_update_display_name(self):
+        # No assertion on the initial value here: asserting `is None` would narrow
+        # the attribute to None for the rest of the function, and mypy would then
+        # read every later comparison as always-false. The default is covered by
+        # test_display_name_defaults_to_none.
```
Comment (jmchilton) [inline test/unit/app/managers/test_UserManager.py:320]: Eliminate this comment and just write a little wrapper assert_user_display_name_is(user, user_name) please. It would improve the readability instead of bringing in specific technologies and such.

**Comment by jmchilton** [comment]
The other AI discovered reuse opportunities codex is rejecting Claude's analysis as unimportant and/or unrelated. I was hoping the comments could be tightened though - see two comments above.

**Review by jmchilton — APPROVED** [review]

File: lib/galaxy_test/api/test_users.py (hunk 1 above)
Comment (itisAliRH) [reply lib/galaxy_test/api/test_users.py]: Agreed, it isn't config dependent. Removed the comment, test unchanged (da3971b7bb).

File: test/unit/app/managers/test_UserManager.py (hunk 2 above)
Comment (itisAliRH) [reply test/unit/app/managers/test_UserManager.py]: Done: added `assert_user_display_name_is(user, display_name)` and used it across the display name manager tests. It also lets the test assert the initial `None` up front, so the comment is gone (fff38b467f).
